src/access/AccessControl.cpp

```diff
@@ -291,6 +291,81 @@ CHIP_ERROR AccessControl::Check(const SubjectDescriptor & subjectDescriptor, con
     return CHIP_ERROR_ACCESS_DENIED;
 }
 
+bool AccessControl::IsValid(const Entry & entry)
+{
+    AuthMode authMode;
+    FabricIndex fabricIndex;
+    Privilege privilege;
+    size_t subjectCount = 0;
+    size_t targetCount  = 0;
+
+    SuccessOrExit(entry.GetAuthMode(authMode));
+    SuccessOrExit(entry.GetFabricIndex(fabricIndex));
+    SuccessOrExit(entry.GetPrivilege(privilege));
+    SuccessOrExit(entry.GetSubjectCount(subjectCount));
+    SuccessOrExit(entry.GetTargetCount(targetCount));
+
+    if (fabricIndex == kUndefinedFabricIndex)
+    {
+        ChipLogError(DataManagement, "AccessControl: invalid fabric index");
+        return false;
+    }
+
+    // Operational PASE not supported for v1.0.
+    if (authMode != AuthMode::kCase && authMode != AuthMode::kGroup)
+    {
+        ChipLogError(DataManagement, "AccessControl: invalid auth mode");
+        return false;
+    }
+
+    if (privilege == Privilege::kAdminister && authMode != AuthMode::kCase)
+    {
+        ChipLogError(DataManagement, "AccessControl: invalid privilege");
+        return false;
+    }
+
+    for (size_t i = 0; i < subjectCount; ++i)
+    {
+        NodeId subject;
+        SuccessOrExit(entry.GetSubject(i, subject));
+        const bool kIsCase  = authMode == AuthMode::kCase;
+        const bool kIsGroup = authMode == AuthMode::kGroup;
+        if ((kIsCase && !IsValidCaseNodeId(subject)) || (kIsGroup && !IsValidGroupNodeId(subject)))
+        {
+            ChipLogError(DataManagement, "AccessControl: invalid subject");
+            return false;
+        }
+    }
+
+    for (size_t i = 0; i < targetCount; ++i)
+    {
+        Entry::Target target;
+        SuccessOrExit(entry.GetTarget(i, target));
+        const bool kHasCluster    = target.flags & Entry::Target::kCluster;
+        const bool kHasEndpoint   = target.flags & Entry::Target::kEndpoint;
+        const bool kHasDeviceType = target.flags & Entry::Target::kDeviceType;
+        if ((!kHasCluster && !kHasEndpoint && !kHasDeviceType) || (kHasEndpoint && kHasDeviceType) ||
```

Comment:
These are all explanatory with the exception of `kHasEndpoint && kHasDeviceType`.
Perhaps add a comment for that one?

Reply:
I simplified this function and flipped the logic on the tests so they could use VerifyOrExit to a common failure code path. Should be clearer now. All the validity checks ultimately come from the spec so that's always the place to look for why things are the way they are.